### PR TITLE
Added pot_vortic module

### PR DESCRIPTION
**Pull request by cfkanesan**
I tried to rebase the original branch but the history was too convoluted and I was afraid of introducing new bugs in the conflict resolution so I had to resort to copying the files over. I also made an attempt at using pytest fixtures to manage the fieldextra run to unload the test body a bit. Let me know what you think.

The tests are currently failing but you can already have a look and comment where relevant.

### CODE REVIEW

File: src/idpi/constants.py
Comment (cfkanesan) [inline src/idpi/constants.py:1]: This was present in Thierry's branch but it seems that the main has diverged from this approach. Should I drop it?

File: src/idpi/operators/curl.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import numpy as np
+
+from .. import constants as const
+
+DEFAULT_DIFF_TYPE = "center"
+STENCILS = {
+    "left": {"m": slice(0, -2), "c": slice(1, -1), "p": slice(1, -1)},
+    "center": {"m": slice(0, -2), "c": slice(1, -1), "p": slice(2, None)},
+    "right": {"m": slice(1, -1), "c": slice(1, -1), "p": slice(2, None)},
+}
```
Comment (cfkanesan) [inline src/idpi/operators/curl.py:11]: I modified the handling of the stencils to be more static and less dependent on the mutation of the global state. That can be a pain to work with in a concurrent execution context.

File: src/idpi/operators/curl.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        - 'm': below (-1)
+        - 'p': above (+1)
+    """
+    if diff_type is None:
+        diff_type = DEFAULT_DIFF_TYPE
+    s = STENCILS[diff_type]
+    return {d: s[k] for d, k in zip("xyz", stencil)}
+
+
+class FDField:
```
Comment (cfkanesan) [inline src/idpi/operators/curl.py:45]: The purpose of this class is to wrap the dataarray and provide syntaxic sugar for the expressions below.

File: src/idpi/operators/curl.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    wk: float,
+    sqrtg_r_s: xr.DataArray,
+    lat: xr.DataArray,
+) -> tuple[xr.DataArray, xr.DataArray, xr.DataArray]:
+    # prepare parameters
+
+    r_earth_inv = 1 / const.earth_radius
+    deg2rad = np.pi / 180
+    acrlat: xr.DataArray = 1 / (np.cos(lat * deg2rad) * const.earth_radius)  # type: ignore[assignment]
+    tgrlat: xr.DataArray = np.tan(lat * deg2rad)  # type: ignore[assignment]
```
Comment (cfkanesan) [inline src/idpi/operators/curl.py:75]: One of the issues in the original branch is that the lattitude values are in degrees and not micro degrees.

File: src/idpi/operators/pot_vortic.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        xr.DataArray: The potential vorticity
+    """
+    stpt = partial(stpt_, diff_type=diff_type) if diff_type else stpt_
+
+    # prepare parameters
+    dlon = U.attrs["GRIB_iDirectionIncrementInDegrees"]
+    dlat = U.attrs["GRIB_jDirectionIncrementInDegrees"]
+    deg2rad = np.pi / 180
+    inv_dlon = 1 / (dlon * deg2rad)
+    inv_dlat = 1 / (dlat * deg2rad)
```
Comment (cfkanesan) [inline src/idpi/operators/pot_vortic.py:69]: The confusion on the units was present here too

File: src/idpi/operators/pot_vortic.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    # FD (finite differences) weights
+    wi = 0.5 * inv_dlon
+    wj = 0.5 * inv_dlat
+    wk = 0.5 if diff_type == "center" else 1.0
+
+    # remove coordinates
+    U, V, W, P, T, HHL, QV, QC, QI = map(
+        lambda a: a.drop_indexes("z"), (U, V, W, P, T, HHL, QV, QC, QI)
+    )
```
Comment (cfkanesan) [inline src/idpi/operators/pot_vortic.py:85]: I'm not the biggest fan of this but at least it doesn't mutate the input variables. I still need to confirm that this works.

File: tests/test_idpi/test_potvortic.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    fs_ds = fieldextra("POT_VORTIC")
+    
+    pv_ref = (
+        fs_ds.rename({"x_1": "x", "y_1": "y", "z_1": "z"})
+        .squeeze(drop=True)
+        .isel(x=slice(1, -1), y=slice(1, -1))  # discard the (undefined) border
+    )
+
+    print(potv.data)
+    print(pv_ref.data)
```
Comment (cfkanesan) [inline tests/test_idpi/test_potvortic.py:55]: leftovers to remove

File: tests/test_idpi/test_potvortic.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import xarray as xr
+import pytest
+
+from idpi import grib_decoder
+import idpi.operators.pot_vortic as pv
+
+
+@pytest.fixture
+def fieldextra_executable(fieldextra_executable):
```
Comment (cfkanesan) [inline tests/test_idpi/test_potvortic.py:10]: Here is an example of the fixture being overridden at the module level.

[] remove this before merging

File: src/idpi/operators/curl.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    dzeta_dphi = (
+        0.25
+        * inv_dlat
+        * sqrtg_r_s
+        * ((hhl["cpc"] - hhl["cmc"]) + (hhl["cpp"] - hhl["cmp"]))
+    )
+
+    # compute weighted derivatives for FD
+
+    du_dphi = ((u["cpc"] + u["mpc"]) - (u["cmc"] + u["mmc"])) * 0.25 * inv_dlat
```
Comment (cosunae) [inline src/idpi/operators/curl.py:109]: I am not very convinced about the usage of this custom syntax. Also notice that sometimes we use it as u["cmc"]  while below we also use u("cmc"). 

I will suggest looking into more standard syntax or aligned with other projects.
* variant 1: 
```u[0,1,0]``` equals the syntax used here as ```u["cpc"]```
it is used also in numba:
https://numba.pydata.org/numba-doc/latest/user/stencil.html
and in gt4py (cartesian mode)
https://github.com/GridTools/gt4py/blob/main/examples/cartesian/demo_horizontal_diffusion.ipynb
* variant 2:
Use a shift operator (concept of gt4py)
```shift(J, 1)(u)``` is equivalent to ```u["cpc"]```
reference: https://hackmd.io/ueLehFeOR5yyOCtn2Qdr-A

* variant 3: 
Use an offset object (used in icon4py)
```u[JOff(1)]``` is equivalent ```u["cpc"]```
reference: 
https://github.com/C2SM/icon4py/blob/main/atm_dyn_iconam/src/icon4py/atm_dyn_iconam/mo_solve_nonhydro_stencil_04.py

**Review by cosunae — CHANGES_REQUESTED** [review]

**Pull request closed without merging**